README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SoilProfileGraphs

<!-- badges: start -->

[![R-CMD-check](https://github.com/brownag/SoilProfileGraphs/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/brownag/SoilProfileGraphs/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of SoilProfileGraphs is to provide extensions for ‘aqp’ and
‘ggplot2’ for creation of soil profile-related graphics. The first focus
is on ‘calibrated’ pit photographs with overplotting of soil property
data.

## Installation

You can install the development version of SoilProfileGraphs like so:

``` r
# install.packages("remotes")
remotes::install_github("brownag/SoilProfileGraphs")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(SoilProfileGraphs)
library(soilDB)

spc <- readRDS(system.file("extdata", "aecpedons.rds",
                           package = "SoilProfileGraphs"))[4, ]
lab <- soilDB::fetchLDM(spc$pedlabsampnum)
#> single result set, returning a data.frame
#> converting profile IDs from integer to character

spg(
  lab,
  system.file("extdata", "sueredo.png",
              package = "SoilProfileGraphs"),
  prop = "sand_total",
  depth_min = -3,
  depth_max = 185,
  no_data_value = 76.5,
  prop_min = 65,
  prop_max = 95
)
#> 2 depth range(s) with missing analytical data removed from 34504.
```

<img src="man/figures/README-example-sueredo-1.jpeg" width="100%" />

``` r
spc <- readRDS(system.file("extdata", "aecpedons.rds",
                           package = "SoilProfileGraphs"))[3, ]
lab <- soilDB::fetchLDM(spc$pedlabsampnum)
#> single result set, returning a data.frame
#> converting profile IDs from integer to character

spg(
  lab,
  system.file("extdata", "buttelake.png",
              package = "SoilProfileGraphs"),
  prop = "sand_total",
  depth_min = -3,
  depth_max = 165,
  no_data_value = 76.5,
  prop_min = 50,
  prop_max = 100
)
#> 1 depth range(s) with missing analytical data removed from 34099.
```

<img src="man/figures/README-example-buttelake-1.jpeg" width="100%" />


NEWS.md
# SoilProfileGraphs 0.0.0.9001

* Added a `NEWS.md` file to track changes to the package.


misc/sueredo.R
library(ggplot2)
library(grid)
library(png)

# load the image
img <- readPNG("inst/extdata/sueredo.png")
spc <- readRDS("inst/extdata/aecpedons.rds")[4, ]
lab <- soilDB::fetchLDM(spc$pedlabsampnum)

depth_min <- -2
depth_max <- 185
prop_min <- 65
prop_max <- 95
prop_breaks <- 6
w <- 1

# data prep
spc <- aqp::glom(spc, depth_min, depth_max)
lab <- aqp::glom(lab, depth_min, depth_max)
a <- aqp::spc2mpspline(lab, "sand_total")
my_data <-
  data.frame(
    x_var = a$sand_total_spline,
    y_var = rowMeans(a[, , .TOP, .BOTTOM])
  )
my_data$x_var[is.na(my_data$x_var)] <- 76.5

# set plot dimensions and aspect ratio
x_axis_limits <- pretty(range(my_data$x_var, na.rm = TRUE),
                        n = 0, min.n = 0)
xidx <- c(which.min(x_axis_limits), which.max(x_axis_limits))
x_axis_limits <- x_axis_limits[xidx]

y_axis_limits <- pretty(range(my_data$y_var, na.rm = TRUE))
yidx <- c(which.min(y_axis_limits), which.max(y_axis_limits))
y_axis_limits <- y_axis_limits[yidx]

img_aspect_ratio <- dim(img)[1] / diff(y_axis_limits) # pixels per vertical cm
dat_aspect_ratio <- dim(img)[2] / diff(x_axis_limits) # pixels per prop unit
plot_ratio <- img_aspect_ratio / dat_aspect_ratio

img_bounds <- mean(x_axis_limits) + (diff(x_axis_limits) / c(-2, 2))
pimg_bounds <- pretty(img_bounds)
pimg_bounds <- pimg_bounds[c(which.min(pimg_bounds), which.max(pimg_bounds))]

p <- ggplot(data = my_data, aes(x = x_var, y = y_var)) +
      theme_minimal() +
      annotation_raster(img, xmin = pimg_bounds[1],
                             xmax = pimg_bounds[2],
                             ymin = -depth_min,
                             ymax = -depth_max) +
      theme(
            panel.ontop = TRUE,
            panel.grid.major = element_line(linetype = "dashed",
                                            color = rgb(1, 1, 1, 0.5)),
            panel.grid.minor = element_blank(),
            axis.ticks.y = element_blank()) +
      geom_path(linetype = "longdash",
                color = "yellow",
                linewidth = 1) +
      scale_x_continuous(limits = c(prop_min, prop_max),
                         n.breaks = prop_breaks) +
      scale_y_reverse() +
      ylim(-y_axis_limits) +
      xlab("Total Sand, %") +
      ylab("Depth, cm") +
      scale_y_reverse() +
      coord_fixed(ratio = plot_ratio)
p
ggsave("man/images/sueredo.jpg", p, device = "jpeg",
       height = 11, width = 8.5, units = "in", dpi = 144)


misc/buttelake.R
library(ggplot2)
library(grid)
library(png)

# Load the image
img <- readPNG("inst/extdata/buttelake.png")
spc <- readRDS("inst/extdata/aecpedons.rds")[3, ]
lab <- soilDB::fetchLDM(spc$pedlabsampnum)

depth_min <- -2
depth_max <- 165
prop_min <- 50
prop_max <- 100
prop_breaks <- 6
w <- 1.25

spc <- trunc(spc, depth_min, depth_max)
lab <- trunc(lab, depth_min, depth_max)
a <- aqp::spc2mpspline(lab, "sand_total")

my_data <-
  data.frame(
    x_var = a$sand_total_spline,
    y_var = rowMeans(a[, , .TOP, .BOTTOM])
  )
my_data$x_var[is.na(my_data$x_var)] <- 76.5

x_axis_limits <- pretty(range(my_data$x_var, na.rm = TRUE),
                        n = 0, min.n = 0)
xidx <- c(which.min(x_axis_limits), which.max(x_axis_limits))
x_axis_limits <- x_axis_limits[xidx]

y_axis_limits <- pretty(range(my_data$y_var, na.rm = TRUE),
                        n = 0, min.n = 0)
yidx <- c(which.min(y_axis_limits), which.max(y_axis_limits))
y_axis_limits <- y_axis_limits[yidx]

img_aspect_ratio <- dim(img)[1] / diff(y_axis_limits) # pixels per vertical cm
dat_aspect_ratio <- dim(img)[2] / diff(x_axis_limits) # pixels per prop unit
plot_ratio <- img_aspect_ratio / dat_aspect_ratio

img_bounds <- mean(x_axis_limits) + (diff(x_axis_limits) / c(-2, 2))
pimg_bounds <- pretty(img_bounds)
pimg_bounds <- pimg_bounds[c(which.min(pimg_bounds), which.max(pimg_bounds))]

p <- ggplot(data = my_data, aes(x = x_var, y = y_var)) +
  theme_minimal() +
  annotation_raster(img, xmin = pimg_bounds[1],
                    xmax = pimg_bounds[2],
                    ymin = -depth_min,
                    ymax = -depth_max) +
  theme(
    panel.ontop = TRUE,
    panel.grid.major = element_line(linetype = "dashed",
                                    color = rgb(1, 1, 1, 0.5)),
    panel.grid.minor = element_blank(),
    axis.ticks.y = element_blank()) +
  geom_path(linetype = "longdash",
            color = "yellow",
            linewidth = 1) +
  scale_x_continuous(limits = c(prop_min, prop_max),
                     n.breaks = prop_breaks) +
  scale_y_reverse() +
  ylim(-y_axis_limits) +
  xlab("Total Sand, %") +
  ylab("Depth, cm") +
  scale_y_reverse() +
  coord_fixed(ratio = plot_ratio)
p
p + annotation_custom(recordPlot())

ggsave("man/images/buttelake.jpg", p, device = "jpeg",
       height = 11, width = 8.5, units = "in", dpi = 144)


R/spg.R
#' spg
#'
#' soil profile property graphs with 'calibrated' pit photo backdrop.
#'
#' @param spc SoilProfileCollection
#' @param filename filename
#' @param prop character. Property column name
#' @param destfile Optional: output filename
#' @param depth_min Minimum image depth (cm)
#' @param depth_max Maximum image depth (cm)
#' @param prop_min Property minimum value
#' @param prop_max Property maximum value
#' @param spline Default: `FALSE`
#' @param midpoint Default: `FALSE` If `spline=FALSE` use midpoints of horizons?
#' @param xlab x-axis label
#' @param ylab y-axis label
#' @param no_data_value Replacement value for `NA` in `prop`
#' @param panel_ontop Default: `TRUE`
#' @param prop_linetype Default: `longdash`
#' @param prop_linewidth Default: `1`
#' @param prop_color Default: `"yellow"`
#' @param prop_breaks Default: `6`
#' @importFrom grDevices rgb
#' @return ggplot plot object; also shown on graphics pane as side-effect.
#' @export
#' @examplesIf !inherits(requireNamespace("soilDB", quietly=TRUE), 'try-error')
#' @examples
#' \donttest{
#' library(soilDB)
#'
#' spc <- readRDS(system.file("extdata", "aecpedons.rds",
#'                    package = "SoilProfileGraphs"))[4, ]
#' lab <- soilDB::fetchLDM(spc$pedlabsampnum)
#'
#' spg(
#'   system.file("extdata",
#'               "sueredo.png",
#'               package = "SoilProfileGraphs"),
#'   spc = lab,
#'   prop = "sand_total",
#'   depth_max = 185,
#'   prop_min = 65,
#'   prop_max = 95
#' )
#' }
spg <-
  function(spc,
           filename,
           prop,
           destfile = NULL,
           depth_min = 0,
           depth_max = 200,
           prop_min = NULL,
           prop_max = NULL,
           spline = FALSE,
           midpoint = FALSE,
           xlab = prop,
           ylab = "Depth, cm",
           no_data_value = NULL,
           panel_ontop = TRUE,
           prop_linetype = "longdash",
           prop_linewidth = 1,
           prop_color = "yellow",
           prop_breaks = 6
  ) {
    .BOTTOM <- NULL
    .TOP <- NULL
    x_var  <- NULL
    y_var  <- NULL

    # Load the image
    img <- png::readPNG(filename)
    spc <- aqp::glom(spc, depth_min, depth_max)

    if (try(requireNamespace("mpspline2", quietly = TRUE)) &&
        (isTRUE(spline) || isFALSE(midpoint))) {
      spc <- aqp::spc2mpspline(spc, prop)
      if (isTRUE(spline)) {
        prop <- paste0(prop, "_spline")
      }
    }

    .data <- data.frame(
      x_var = spc[[prop]],
      y_var = rowMeans(spc[, , .TOP, .BOTTOM])
    )

  if (!is.null(no_data_value))
    .data$x_var[is.na(.data$x_var)] <- no_data_value

  xlim <- pretty(range(.data$x_var, na.rm = TRUE),
                          n = 0, min.n = 0)
  xidx <- c(which.min(xlim), which.max(xlim))
  xlim <- xlim[xidx]

  ylim <- pretty(range(.data$y_var, na.rm = TRUE),
                          n = 0, min.n = 0)
  yidx <- c(which.min(ylim), which.max(ylim))
  ylim <- ylim[yidx]

  img_aspect_ratio <- dim(img)[2] / dim(img)[1]
  dat_aspect_ratio <-  diff(xlim) / diff(ylim)
  plot_ratio <- mean(c(img_aspect_ratio, dat_aspect_ratio))

  img_bounds <- mean(xlim) + (diff(xlim) / c(-2, 2))
  pimg_bounds <- pretty(img_bounds)
  pimg_bounds <- pimg_bounds[c(which.min(pimg_bounds), which.max(pimg_bounds))]

  suppressMessages({
    p <- ggplot2::ggplot(data = .data, ggplot2::aes(x = x_var, y = y_var)) +
      ggplot2::theme_minimal() +
      ggplot2::annotation_raster(img, xmin = pimg_bounds[1],
                        xmax = pimg_bounds[2],
                        ymin = -depth_min,
                        ymax = -depth_max) +
      ggplot2::theme(
        panel.ontop = TRUE,
        panel.grid.major = ggplot2::element_line(linetype = "dashed",
                                                 color = grDevices::rgb(1, 1, 1, 0.5)),
        panel.grid.minor = ggplot2::element_blank(),
        axis.ticks.y = ggplot2::element_blank()
      ) +
      ggplot2::geom_path(linetype = prop_linetype,
                         color = prop_color,
                         linewidth = prop_linewidth) +
      ggplot2::scale_y_reverse() +
      ggplot2::scale_x_continuous(limits = c(prop_min, prop_max),
                                  n.breaks = prop_breaks) +
      ggplot2::ylim(-ylim) +
      ggplot2::xlab(xlab) +
      ggplot2::ylab(ylab) +
      ggplot2::scale_y_reverse() +
      ggplot2::coord_fixed(ratio = plot_ratio)
  })

  # TODO: expose file output options
  if (!is.null(destfile)) {
    ggplot2::ggsave(destfile, p, device = "jpeg",
           height = 11, width = 8.5, units = "in", dpi = 144)
  }
  p
}
